dynamic_validators/insecure_dynamic_validators.md

```diff
@@ -0,0 +1,103 @@
+
+- `GET /api/v1/validators` should be introduced to list the current validators.
+
+### Technical challenges
+Since bdb clients can call `PUT /api/v1/validators` anytime it is necessary to (temporarily) store the payload until `end_block` is called. This implies that a successful `PUT` doesn't necessarily mean that the validators list has been updated. Moreover, one also needs to consider a scenario wherein the Tendermint node goes offline for whatever reason. When the node is brought back online it would load the validators list from its config file which may not include dynamically added validators.
+
+Furthermore, as discussed in [technical details](#technical-details) each/majority (>2/3) nodes in the network need to add the new node to their list of validators in order for the new node to be able to act as a validator. Until then the node may not be able to participate and vote on blocks.
+
+
+### Storing dynamically added validators
+As discussed in the previous section it is necessary to (temporarily) store the updates to validators list until `end_block` is executed. Furthermore, in case of a node crash it is required that that upon restart the node should be able to recover its dynamically added validators. Following is one possible solution,
+
+- Store the validators in a MongoDB collection along with a flag i.e.,
+```json
+{"validators": [],
+ "sync": true
+}
+```
+- The sync flag is used during `end_block` to decide if a validator update should be returned.
+- The sync flag is set to `True` when api `PUT /api/v1/validators` is called and the `"validators"` are updated.
+- Whenever the sync flag is set, `end_block` will call `http://tendermint:46657/validators`, calculate the diff with the list of validators retrieved from mongodb and return it.
+- When bigchaindb starts the sync flag should be set to `True`.
+
+
+### API impact
+New api endpoint `PUT /api/v1/validators` will introduced,
+
+- Add/remove validators
+  - Uri: `PUT /api/v1/validators`
+  - Body: 
+  ```json
+  [{"pub_key":{"type":"ed25519","data":"4E2685D9016126864733225BE00F005515200727FBAB1312FC78C8B76831255A"},
+    "power":10},
+   {"pub_key":{"type":"ed25519","data":"608D839D7100466D6BA6BE79C320F8B81DE93CFAA58CF9768CF921C6371F2553"},
+   "power":0}
+   ]
+   ```
+   - response: `200`
+
+- List validators
+  - Uri: `GET /api/v1/validators`
+  - response: 
+  ```json
+  [{"pub_key":{"type":"ed25519","data":"4E2685D9016126864733225BE00F005515200727FBAB1312FC78C8B76831255A"},
+    "power":10},
+   {"pub_key":{"type":"ed25519","data":"608D839D7100466D6BA6BE79C320F8B81DE93CFAA58CF9768CF921C6371F2553"},
+   "power":0}
+   ]
+  ```
+
+**NOTE**: when `power: 0` is set for a given validator it is removed from the validators list.
+
+### Security impact
+The current implementation is insecure i.e. any user who gains access to a given node can modify its validators. This node could effectively be considered byzantine and the network can tolerate (<1/3) byzantine nodes.
```

Comment:
It's not clear to me why this is insecure for the overall system. Here's why:

Suppose a BigchainDB node operator uses the `end_block` response to tell his local Tendermint instance to update its local list of validators. It will do that, but only locally, not on any other Tendermint instance. The "damage" is limited to just that node.

If enough node operators add the same new validator to their list (i.e. with more than 2/3 of the total power), then the new validator will be recognized as a legitimate validator and its votes will be accepted, otherwise its votes will be ignored. Is that right? If so, that seems fairly secure to me.